src/ucp/rndv/rndv_get.c

```diff
@@ -45,11 +43,112 @@ ucp_proto_rndv_get_zcopy_init(const ucp_proto_init_params_t *init_params)
+                            req->send.rndv.remote_address +
+                                    req->send.state.dt_iter.offset,
+                            tl_rkey, &req->send.state.uct_comp);
+}
+
+static ucs_status_t ucp_proto_rndv_get_zcopy_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_rndv_bulk_priv_t *rpriv = req->send.proto_config->priv;
+
+    if (ucp_datatype_iter_is_end(&req->send.state.dt_iter)) {
+        if (req->send.state.dt_iter.length > 0) {
+            ucs_assert(req->send.state.uct_comp.count == 0);
+        }
+        return ucp_proto_rndv_ack_progress(req, UCP_AM_ID_RNDV_ATS,
+                                           ucp_proto_rndv_get_complete);
+    } else {
+        return ucp_proto_multi_zcopy_progress(req, &rpriv->mpriv, NULL,
+                                              ucp_proto_rndv_get_zcopy_send_func,
+                                              ucp_proto_rndv_get_completion);
+    }
+}
+
 static ucp_proto_t ucp_rndv_get_zcopy_proto = {
     .name       = "rndv/get/zcopy",
     .flags      = 0,
     .init       = ucp_proto_rndv_get_zcopy_init,
     .config_str = ucp_proto_rndv_bulk_config_str,
-    .progress   = (uct_pending_callback_t)ucs_empty_function_do_assert
+    .progress   = ucp_proto_rndv_get_zcopy_progress
 };
 UCP_PROTO_REGISTER(&ucp_rndv_get_zcopy_proto);
+
+
+static ucs_status_t
+ucp_proto_rndv_ats_init(const ucp_proto_init_params_t *params)
+{
+    ucs_status_t status;
+
+    if (params->select_param->op_id != UCP_OP_ID_RNDV_RECV) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (params->rkey_config_key != NULL) {
+        /* This ATS-only protocol will not take care of releasing the remote, so
+           disqualify if remote key is present */
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    status = ucp_proto_rndv_ack_init(params);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* Support only 0-length messages */
+    *params->priv_size                 = sizeof(ucp_proto_rndv_ack_priv_t);
+    params->caps->cfg_thresh           = 0;
+    params->caps->cfg_priority         = 1;
+    params->caps->min_length           = 0;
+    params->caps->num_ranges           = 1;
+    params->caps->ranges[0].max_length = 0;
+    params->caps->ranges[0].perf       = ucp_proto_rndv_ack_time(params);
+    return UCS_OK;
+}
+
+static ucs_status_t ucp_proto_rndv_ats_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    return ucp_proto_rndv_ack_progress(
+            req, UCP_AM_ID_RNDV_ATS, ucp_proto_request_zcopy_complete_success);
+}
+
+static ucp_proto_t ucp_rndv_ats_proto = {
+    .name       = "rndv/ats",
+    .flags      = 0,
+    .init       = ucp_proto_rndv_ats_init,
+    .config_str = ucp_proto_rndv_ack_config_str,
+    .progress   = ucp_proto_rndv_ats_progress
+};
```

Comment:
do we really need to define a separate proto for ack?

Reply:
currently get_zcopy requires local registration and remote key. however for 0-length messages we don't need them, on other hand it's not possible that a protocol would require different things for different message sizes. But it's possible to define a message size range a protocol can handle.